File: my_flutter_app/functions/test/invoice-ledger.test.js
"use strict";

const assert = require("node:assert/strict");
const {describe, it} = require("node:test");

const {
  validateInvoice,
  invoiceRecord,
  validateInvoiceLine,
  invoiceLineRecord,
  validateInvoicePayment,
  invoicePaymentRecord,
  invoiceTotals,
  invoiceIsOverdue,
  invoiceNumber,
  invoiceKind,
  invoiceTextSummary,
  dayKey,
} = require("../invoice_ledger");

describe("an invoice", () => {
  it("needs a title and a customer; dates must be real and in order", () => {
    assert.deepEqual(validateInvoice({}),
        ["title_required", "customer_name_required"]);
    assert.deepEqual(
        validateInvoice({title: "Corolla", customerName: "Amadou"}), []);
    assert.deepEqual(validateInvoice({title: "x", customerName: "y",
      issuedOn: "2026-02-30"}), ["issued_on_invalid"]);
    assert.deepEqual(validateInvoice({title: "x", customerName: "y",
      issuedOn: "2026-09-20", dueOn: "2026-09-01"}), ["due_before_issued"]);
  });

  it("stores calendar days, defaulting the invoice date to today", () => {
    const rec = invoiceRecord({title: " Corolla ", customerName: "Amadou Bah",
      customerPhone: "+1 646", dueOn: "2026-10-15T12:00:00"}, "2026-09-26");
    assert.equal(rec.title, "Corolla");
    assert.equal(rec.issuedOn, "2026-09-26");
    assert.equal(rec.dueOn, "2026-10-15");
    assert.equal(dayKey({toDate: () => new Date("2026-09-01T23:00:00Z")}),
        "2026-09-01");
  });

  it("numbers per business, four digits, never reused", () => {
    assert.equal(invoiceNumber(1), "INV-0001");
    assert.equal(invoiceNumber(12345), "INV-12345");
  });
});

describe("a line", () => {
  it("is a description, a count and a price for one; amount derived", () => {
    assert.deepEqual(validateInvoiceLine({}),
        ["description_required", "quantity_required", "unit_price_invalid"]);
    assert.deepEqual(validateInvoiceLine({description: "Barrels", quantity: 8,
      unitPriceCents: 12000}), []);
    const line = invoiceLineRecord({description: " Barrels to Conakry ",
      quantity: "8", unitPriceCents: 12000, vinNumber: ""});
    assert.equal(line.amountCents, 96000);
    assert.equal(line.quantity, 8);
  });

  // A car line carries the VIN on the paper; a half-typed one is refused
  // rather than printed wrong.
  it("keeps a VIN only when it is a whole one", () => {
    assert.deepEqual(validateInvoiceLine({description: "Corolla", quantity: 1,
      unitPriceCents: 550000, vinNumber: "1hgcm82633a004352"}), []);
    assert.deepEqual(validateInvoiceLine({description: "Corolla", quantity: 1,
      unitPriceCents: 550000, vinNumber: "ABC123"}), ["vin_invalid"]);
    assert.equal(invoiceLineRecord({description: "Corolla", quantity: 1,
      unitPriceCents: 1, vinNumber: " 1hgcm82633a004352 "}).vinNumber,
    "1HGCM82633A004352");
  });

  it("refuses a typo-sized amount", () => {
    assert.deepEqual(validateInvoiceLine({description: "x", quantity: 10000,
      unitPriceCents: 10001}), ["amount_too_large"]);
  });
});

describe("a payment", () => {
  it("must be positive, within the balance, and say how it arrived", () => {
    assert.deepEqual(validateInvoicePayment({}, 10000),
        ["amount_required", "payment_method_invalid"]);
    assert.deepEqual(validateInvoicePayment(
        {amountCents: 20000, method: "cash"}, 10000),
    ["payment_exceeds_balance"]);
    assert.deepEqual(validateInvoicePayment(
        {amountCents: 10000, method: "cash"}, 10000), []);
    assert.deepEqual(validateInvoicePayment({amountCents: 100, method: "wire"},
        10000), ["payment_method_invalid"]);
    const rec = invoicePaymentRecord({amountCents: 5000, method: "zelle"},
        "2026-09-26");
    assert.equal(rec.paidOn, "2026-09-26");
    assert.equal(rec.reverted, false);
  });
});

describe("what the invoice adds up to", () => {
  const lines = [{amountCents: 550000}, {amountCents: 5000}];
  it("is total minus the payments that still stand", () => {
    const t = invoiceTotals(lines, [
      {amountCents: 200000}, {amountCents: 100000, reverted: true}]);
    assert.equal(t.totalCents, 555000);
    assert.equal(t.paidCents, 200000);
    assert.equal(t.balanceCents, 355000);
    assert.equal(t.paymentCount, 1);
    assert.equal(t.status, "open");
  });

  it("is paid when nothing is owed; a blank invoice is never paid", () => {
    assert.equal(invoiceTotals(lines, [{amountCents: 555000}]).status, "paid");
    assert.equal(invoiceTotals([], []).status, "open");
    assert.equal(invoiceKind({status: "paid"}), "Receipt");
    assert.equal(invoiceKind({status: "open"}), "Invoice");
  });

  it("is overdue only while open and past its due day", () => {
    assert.equal(invoiceIsOverdue({status: "open", dueOn: "2026-09-20"},
        "2026-09-26"), true);
    assert.equal(invoiceIsOverdue({status: "open", dueOn: "2026-09-26"},
        "2026-09-26"), false);
    assert.equal(invoiceIsOverdue({status: "paid", dueOn: "2026-09-20"},
        "2026-09-26"), false);
    assert.equal(invoiceIsOverdue({status: "open"}, "2026-09-26"), false);
  });
});

describe("the WhatsApp text", () => {
  it("reads the lines, the total, what was paid and what is owed", () => {
    const txt = invoiceTextSummary({
      businessName: "Keren Auto Sales",
      invoice: {number: "INV-0007", title: "Corolla",
        customerName: "Amadou Bah", issuedOn: "2026-09-26",
        dueOn: "2026-10-15"},
      lines: [
        {description: "2014 Toyota Corolla", quantity: 1,
          unitPriceCents: 550000, amountCents: 550000,
          vinNumber: "1HGCM82633A004352"},
        {description: "Barrels", quantity: 8, unitPriceCents: 12000,
          amountCents: 96000},
      ],
      payments: [{amountCents: 200000}],
    });
    assert.match(txt, /^Keren Auto Sales — Invoice INV-0007\n/);
    assert.match(txt,
        /2014 Toyota Corolla — \$5,500\.00\nVIN 1HGCM82633A004352/);
    assert.match(txt, /Barrels \(8 × \$120\.00\) — \$960\.00/);
    assert.match(txt, /Total: \$6,460\.00\nPaid: -\$2,000\.00\n/);
    assert.match(txt, /BALANCE DUE: \$4,460\.00 \(due 2026-10-15\)$/);
    const paid = invoiceTextSummary({businessName: "K", invoice: {title: "t",
      customerName: "c"}, lines: [{amountCents: 100}],
    payments: [{amountCents: 100}]});
    assert.match(paid, /— Receipt\n/);
    assert.match(paid, /PAID IN FULL$/);
  });
});

File: my_flutter_app/test/invoice_ledger_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:my_flutter_app/services/invoice_ledger.dart';

Invoice inv({
  String id = 'i1',
  String number = 'INV-0007',
  String title = 'Corolla',
  String customer = 'Amadou Bah',
  String status = invoiceStatusOpen,
  String issuedOn = '2026-09-26',
  String dueOn = '2026-10-15',
}) =>
    Invoice.fromMap(id, {
      'businessId': 'b1',
      'number': number,
      'title': title,
      'customerName': customer,
      'status': status,
      'issuedOn': issuedOn,
      'dueOn': dueOn,
    });

void main() {
  group('an invoice', () {
    test('needs a title and a customer; dates must be real and in order', () {
      expect(validateInvoice(const InvoiceDraft()),
          ['title_required', 'customer_name_required']);
      expect(
          validateInvoice(
              const InvoiceDraft(title: 'Corolla', customerName: 'Amadou')),
          isEmpty);
      expect(
          validateInvoice(const InvoiceDraft(
              title: 'x', customerName: 'y', issuedOn: '2026-02-30')),
          ['issued_on_invalid']);
      expect(
          validateInvoice(const InvoiceDraft(
              title: 'x',
              customerName: 'y',
              issuedOn: '2026-09-20',
              dueOn: '2026-09-01')),
          ['due_before_issued']);
    });

    test('dates are calendar days', () {
      expect(invoiceDayKey('2026-10-15T12:00:00'), '2026-10-15');
      expect(invoiceDayKey(DateTime.utc(2026, 9, 1, 23)), '2026-09-01');
      expect(invoiceDayKey(''), '');
    });
  });

  group('a line', () {
    test('is typed in dollars and sent in cents, amount worked out server-side',
        () {
      expect(invoiceDollarsToCents('1,250.50'), 125050);
      expect(invoiceDollarsToCents(r'$120'), 12000);
      expect(invoiceDollarsToCents('abc'), isNull);
      expect(invoiceCentsToInput(12000), '120');
      expect(invoiceCentsToInput(12050), '120.50');
      expect(validateInvoiceLine(const InvoiceLineDraft()),
          ['description_required', 'unit_price_invalid']);
      expect(
          invoiceLinePayload(const InvoiceLineDraft(
              description: ' Barrels ', quantity: 8, unitPriceCents: 12000)),
          {
            'description': 'Barrels',
            'quantity': 8,
            'unitPriceCents': 12000,
            'vinNumber': ''
          });
    });

    test('keeps a VIN only when it is a whole one', () {
      expect(
          validateInvoiceLine(const InvoiceLineDraft(
              description: 'Corolla',
              unitPriceCents: 550000,
              vinNumber: '1hgcm82633a004352')),
          isEmpty);
      expect(
          validateInvoiceLine(const InvoiceLineDraft(
              description: 'Corolla', unitPriceCents: 1, vinNumber: 'ABC')),
          ['vin_invalid']);
      expect(
          invoiceLinePayload(const InvoiceLineDraft(
              description: 'c',
              unitPriceCents: 1,
              vinNumber: ' 1hgcm82633a004352 '))['vinNumber'],
          '1HGCM82633A004352');
    });
  });

  group('a payment', () {
    test('must fit the balance and say how it arrived', () {
      expect(validateInvoicePayment(const InvoicePaymentDraft(), 10000),
          ['amount_required']);
      expect(
          validateInvoicePayment(
              const InvoicePaymentDraft(amountCents: 20000), 10000),
          ['payment_exceeds_balance']);
      expect(
          validateInvoicePayment(
              const InvoicePaymentDraft(amountCents: 100, method: 'wire'),
              10000),
          ['payment_method_invalid']);
      expect(
          validateInvoicePayment(
              const InvoicePaymentDraft(amountCents: 10000, method: 'zelle'),
              10000),
          isEmpty);
    });
  });

  group('what the invoice adds up to', () {
    final lines = [
      InvoiceLine.fromMap('l1', {'amountCents': 550000}),
      InvoiceLine.fromMap('l2', {'amountCents': 5000}),
    ];
    test('is total minus the payments that still stand', () {
      final t = invoiceTotals(lines, [
        InvoicePayment.fromMap('p1', {'amountCents': 200000}),
        InvoicePayment.fromMap('p2', {'amountCents': 100000, 'reverted': true}),
      ]);
      expect(t.totalCents, 555000);
      expect(t.paidCents, 200000);
      expect(t.balanceCents, 355000);
      expect(t.paymentCount, 1);
      expect(t.status, invoiceStatusOpen);
    });

    test('is paid when nothing is owed; a blank invoice is never paid', () {
      expect(
          invoiceTotals(
              lines, [InvoicePayment.fromMap('p', {'amountCents': 555000})])
              .status,
          invoiceStatusPaid);
      expect(invoiceTotals(const [], const []).status, invoiceStatusOpen);
    });

    test('overdue only while open and past the due day', () {
      expect(invoiceIsOverdue(inv(dueOn: '2026-09-20'), '2026-09-26'), isTrue);
      expect(invoiceIsOverdue(inv(dueOn: '2026-09-26'), '2026-09-26'), isFalse);
      expect(
          invoiceIsOverdue(
              inv(status: invoiceStatusPaid, dueOn: '2026-09-20'), '2026-09-26'),
          isFalse);
      expect(invoiceIsOverdue(inv(dueOn: ''), '2026-09-26'), isFalse);
    });

    test('the list filters and sorts like the console', () {
      final rows = [
        inv(id: 'a', number: 'INV-0001', issuedOn: '2026-09-01', dueOn: '2026-09-20'),
        inv(id: 'b', number: 'INV-0002', title: 'Barrels', customer: 'Fatou',
            status: invoiceStatusPaid, issuedOn: '2026-09-10', dueOn: ''),
        inv(id: 'c', number: 'INV-0003', title: 'Tyres', issuedOn: '2026-09-10', dueOn: ''),
      ];
      List<String> ids(String f, String q) =>
          filterInvoices(rows, f, q, '2026-09-26').map((r) => r.id).toList();
      expect(ids(invoiceFilterOpen, ''), ['a', 'c']);
      expect(ids(invoiceFilterPaid, ''), ['b']);
      expect(ids(invoiceFilterOverdue, ''), ['a']);
      expect(ids(invoiceFilterAll, 'amadou'), ['a', 'c']);
      expect(ids(invoiceFilterAll, '0002'), ['b']);
      expect(sortInvoices(rows).map((r) => r.id).toList(), ['c', 'b', 'a']);
    });
  });

  test('the WhatsApp text mirrors the server byte for byte', () {
    final txt = invoiceTextSummary(
      businessName: 'Keren Auto Sales',
      invoice: inv(),
      lines: [
        InvoiceLine.fromMap('l1', {
          'description': '2014 Toyota Corolla',
          'quantity': 1,
          'unitPriceCents': 550000,
          'amountCents': 550000,
          'vinNumber': '1HGCM82633A004352',
        }),
        InvoiceLine.fromMap('l2', {
          'description': 'Barrels',
          'quantity': 8,
          'unitPriceCents': 12000,
          'amountCents': 96000,
        }),
      ],
      payments: [
        InvoicePayment.fromMap('p1', {
          'amountCents': 200000,
          'paidOn': '2026-09-26',
          'method': 'zelle',
          'forDescription': '2014 Toyota Corolla',
        })
      ],
    );
    expect(txt, [
      'Keren Auto Sales — Invoice INV-0007',
      'Corolla · 2026-09-26',
      'For: Amadou Bah',
      '',
      '2014 Toyota Corolla — \$5,500.00\nVIN 1HGCM82633A004352',
      'Barrels (8 × \$120.00) — \$960.00',
      '',
      'Total: \$6,460.00',
      'Paid: -\$2,000.00',
      '  2026-09-26 · zelle · for 2014 Toyota Corolla — \$2,000.00',
      'BALANCE DUE: \$4,460.00 (due 2026-10-15)',
    ].join('\n'));
  });
}
